File: bbike/app/api/v1/endpoints/analytics.py
from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session
from sqlalchemy import func, desc
from app.core.database import get_db
from app.models.snapshot import BikeSnapshot, StationSnapshot
from datetime import datetime, timedelta, timezone
from typing import List

router = APIRouter()

@router.get("/snapshots/latest")
async def get_latest_snapshot(db: Session = Depends(get_db)):
    """Get the most recent snapshot data"""
    
    latest_bike_time = db.query(func.max(BikeSnapshot.timestamp)).scalar()
    latest_station_time = db.query(func.max(StationSnapshot.timestamp)).scalar()
    
    if not latest_bike_time or not latest_station_time:
        return {"error": "No snapshots available yet"}
    
    bikes = db.query(BikeSnapshot).filter(
        BikeSnapshot.timestamp == latest_bike_time
    ).all()
    
    stations = db.query(StationSnapshot).filter(
        StationSnapshot.timestamp == latest_station_time
    ).all()
    
    return {
        "timestamp": latest_bike_time,
        "bikes": {
            "total": len(bikes),
            "available": len([b for b in bikes if not b.is_reserved and not b.is_disabled]),
            "reserved": len([b for b in bikes if b.is_reserved]),
            "disabled": len([b for b in bikes if b.is_disabled])
        },
        "stations": {
            "total": len(stations),
            "active": len([s for s in stations if s.is_installed and s.is_renting]),
            "total_bikes": sum(s.num_bikes_available for s in stations),
            "total_ebikes": sum(s.num_ebikes_available for s in stations)
        }
    }

@router.get("/snapshots/history")
async def get_snapshot_history(
    hours: int = 24,
    db: Session = Depends(get_db)
):
    """Get snapshot history for the last N hours"""
    
    cutoff_time = datetime.now(timezone.utc) - timedelta(hours=hours)
    
    # Get unique timestamps
    timestamps = db.query(BikeSnapshot.timestamp).filter(
        BikeSnapshot.timestamp >= cutoff_time
    ).distinct().order_by(desc(BikeSnapshot.timestamp)).all()
    
    history = []
    for (ts,) in timestamps:
        bikes = db.query(BikeSnapshot).filter(
            BikeSnapshot.timestamp == ts,
            BikeSnapshot.is_reserved == False,
            BikeSnapshot.is_disabled == False
        ).count()
        
        stations_active = db.query(StationSnapshot).filter(
            StationSnapshot.timestamp == ts,
            StationSnapshot.is_renting == True
        ).count()
        
        total_bikes_at_stations = db.query(
            func.sum(StationSnapshot.num_bikes_available)
        ).filter(
            StationSnapshot.timestamp == ts
        ).scalar() or 0
        
        history.append({
            "timestamp": ts,
            "bikes_available": bikes,
            "stations_active": stations_active,
            "total_bikes_at_stations": total_bikes_at_stations
        })
    
    return {"hours": hours, "data_points": len(history), "history": history}

@router.get("/stations/popular")
async def get_popular_stations(
    limit: int = 10,
    db: Session = Depends(get_db)
):
    """Get most popular stations by average bike availability"""
    
    latest_time = db.query(func.max(StationSnapshot.timestamp)).scalar()
    
    if not latest_time:
        return {"error": "No data available"}
    
    # Get stations from last 24 hours
    cutoff = datetime.now(timezone.utc) - timedelta(hours=24)
    
    popular = db.query(
        StationSnapshot.station_id,
        StationSnapshot.name,
        func.avg(StationSnapshot.num_bikes_available).label('avg_bikes')
    ).filter(
        StationSnapshot.timestamp >= cutoff
    ).group_by(
        StationSnapshot.station_id,
        StationSnapshot.name
    ).order_by(
        desc('avg_bikes')
    ).limit(limit).all()
    
    return {
        "popular_stations": [
            {
                "station_id": s.station_id,
                "name": s.name,
                "average_bikes": round(s.avg_bikes, 2)
            }
            for s in popular
        ]
    }

File: worker/src/routes/analytics.ts
import { Hono } from "hono";
import type { Env } from "../index";
import { resolveCity } from "../cities";

// Ported from bbike/app/api/v1/endpoints/analytics.py

const analytics = new Hono<{ Bindings: Env }>();

analytics.get("/snapshots/latest", async (c) => {
  const city = resolveCity(c.req.query("city"));
  if (!city) return c.json({ detail: "Unknown city" }, 400);
  const [bikeTs, stationTs] = await Promise.all([
    c.env.DB
      .prepare("SELECT MAX(timestamp) AS ts FROM bike_snapshots WHERE city = ?")
      .bind(city)
      .first<string>("ts"),
    c.env.DB
      .prepare("SELECT MAX(timestamp) AS ts FROM station_snapshots WHERE city = ?")
      .bind(city)
      .first<string>("ts"),
  ]);
  // Python returns a 200 with an error body here, not a 5xx
  if (!bikeTs || !stationTs) return c.json({ error: "No snapshots available yet" });

  const [bikeAgg, stationAgg] = await c.env.DB.batch([
    c.env.DB
      .prepare(
        `SELECT COUNT(*) AS total,
                SUM(CASE WHEN is_reserved = 0 AND is_disabled = 0 THEN 1 ELSE 0 END) AS available,
                SUM(is_reserved) AS reserved,
                SUM(is_disabled) AS disabled
         FROM bike_snapshots WHERE city = ? AND timestamp = ?`
      )
      .bind(city, bikeTs),
    c.env.DB
      .prepare(
        `SELECT COUNT(*) AS total,
                SUM(CASE WHEN is_installed = 1 AND is_renting = 1 THEN 1 ELSE 0 END) AS active,
                SUM(num_bikes_available) AS total_bikes,
                SUM(num_ebikes_available) AS total_ebikes
         FROM station_snapshots WHERE city = ? AND timestamp = ?`
      )
      .bind(city, stationTs),
  ]);

  const b = bikeAgg.results[0] as Record<string, number>;
  const s = stationAgg.results[0] as Record<string, number>;
  return c.json({
    timestamp: bikeTs,
    bikes: {
      total: b.total,
      available: b.available ?? 0,
      reserved: b.reserved ?? 0,
      disabled: b.disabled ?? 0,
    },
    stations: {
      total: s.total,
      active: s.active ?? 0,
      total_bikes: s.total_bikes ?? 0,
      total_ebikes: s.total_ebikes ?? 0,
    },
  });
});

analytics.get("/snapshots/history", async (c) => {
  const city = resolveCity(c.req.query("city"));
  if (!city) return c.json({ detail: "Unknown city" }, 400);
  const hours = Number(c.req.query("hours") ?? 24);
  const cutoff = new Date(Date.now() - hours * 3600_000).toISOString();

  // The Python version ran 3 queries per timestamp (N+1); this does one GROUP BY per table
  const [bikeRows, stationRows] = await c.env.DB.batch([
    c.env.DB
      .prepare(
        `SELECT timestamp,
                SUM(CASE WHEN is_reserved = 0 AND is_disabled = 0 THEN 1 ELSE 0 END) AS bikes_available
         FROM bike_snapshots WHERE city = ? AND timestamp >= ?
         GROUP BY timestamp ORDER BY timestamp DESC`
      )
      .bind(city, cutoff),
    c.env.DB
      .prepare(
        `SELECT timestamp,
                SUM(is_renting) AS stations_active,
                SUM(num_bikes_available) AS total_bikes_at_stations
         FROM station_snapshots WHERE city = ? AND timestamp >= ?
         GROUP BY timestamp`
      )
      .bind(city, cutoff),
  ]);

  const stationList = stationRows.results as {
    timestamp: string;
    stations_active: number;
    total_bikes_at_stations: number;
  }[];

  // Bike and station snapshots for a city share a timestamp within a cron tick,
  // but tolerate drift by pairing each bike row with the nearest earlier station row
  const stationByTs = new Map(stationList.map((r) => [r.timestamp, r]));
  const history = (bikeRows.results as { timestamp: string; bikes_available: number }[]).map((r) => ({
    timestamp: r.timestamp,
    bikes_available: r.bikes_available,
    stations_active: stationByTs.get(r.timestamp)?.stations_active ?? 0,
    total_bikes_at_stations: stationByTs.get(r.timestamp)?.total_bikes_at_stations ?? 0,
  }));

  return c.json({ hours, data_points: history.length, history });
});

analytics.get("/stations/popular", async (c) => {
  const city = resolveCity(c.req.query("city"));
  if (!city) return c.json({ detail: "Unknown city" }, 400);
  const limit = Number(c.req.query("limit") ?? 10);
  const hasData = await c.env.DB
    .prepare("SELECT MAX(timestamp) AS ts FROM station_snapshots WHERE city = ?")
    .bind(city)
    .first<string>("ts");
  if (!hasData) return c.json({ error: "No data available" });

  const cutoff = new Date(Date.now() - 24 * 3600_000).toISOString();
  const { results } = await c.env.DB
    .prepare(
      `SELECT station_id, name, AVG(num_bikes_available) AS avg_bikes
       FROM station_snapshots WHERE city = ? AND timestamp >= ?
       GROUP BY station_id, name
       ORDER BY avg_bikes DESC LIMIT ?`
    )
    .bind(city, cutoff, limit)
    .all<{ station_id: string; name: string; avg_bikes: number }>();

  return c.json({
    popular_stations: results.map((s) => ({
      station_id: s.station_id,
      name: s.name,
      average_bikes: Math.round(s.avg_bikes * 100) / 100,
    })),
  });
});

export default analytics;
